realm/realm-jni/src/io_realm_internal_TableQuery.cpp

```diff
@@ -1037,23 +1037,23 @@ JNIEXPORT jlong JNICALL Java_io_realm_internal_TableQuery_nativeFind(
     return -1;
 }
 
-std::unique_ptr<Query> getHandoverQuery (jlong bgSharedGroupPtr, jlong queryPtr, bool advanceToLatestVersion)
+std::unique_ptr<Query> handoverQueryToWorker(jlong bgSharedGroupPtr, jlong queryPtr, bool advanceToLatestVersion)
 {
     SharedGroup::Handover<Query> *handoverQueryPtr = HO(Query, queryPtr);
     std::unique_ptr<SharedGroup::Handover<Query>> handoverQuery(handoverQueryPtr);
 
-    SG(bgSharedGroupPtr)->end_read();
+    // if the SharedGroup is not in Read Transaction, we position it at the same version as the handover
+    if (SG(bgSharedGroupPtr)->get_transact_stage() != SharedGroup::transact_Reading) {
+        SG(bgSharedGroupPtr)->begin_read(handoverQuery->version);
 
-    SharedGroup::VersionID currentVersion = SG(bgSharedGroupPtr)->get_version_of_current_transaction();
-    bool isDifferentVersions = (currentVersion != handoverQuery->version);
-    if (isDifferentVersions) {
+    } else if (SG(bgSharedGroupPtr)->get_version_of_current_transaction() != handoverQuery->version) {
+        SG(bgSharedGroupPtr)->end_read();
         SG(bgSharedGroupPtr)->begin_read(handoverQuery->version);
```

Comment:
This line can still cause a BadVersion which is currently not getting caught and will crash. 
Also, right now our executor service has more than 1 thread, by restricting it to 1 thread It seems we can control the concurrency enough for a stable unit tests.

I have an idea for fixing it + actually making a proper unit test for this. Update incoming